Libplanet/Blockchain/Renderers/DelayedRenderer.cs

```diff
@@ -0,0 +1,293 @@
+        /// if the tip change is a reorg.  Otherwise returns <c>null</c>.</returns>
+        protected virtual Block<T>? OnTipChanged(Block<T>? oldTip, Block<T> newTip)
+        {
+            if (oldTip is null)
+            {
+                return null;
+            }
+
+            Block<T>? branchpoint = null;
+            if (!newTip.PreviousHash.Equals(oldTip.Hash))
+            {
+                branchpoint = FindBranchpoint(oldTip, newTip);
+                Renderer.RenderReorg(oldTip, newTip, branchpoint);
+            }
+
+            Renderer.RenderBlock(oldTip, newTip);
+            return branchpoint;
+        }
+
+        private void DiscoverBlock(Block<T> block)
+        {
+            if (_confirmed.ContainsKey(block.Hash))
+            {
+                return;
+            }
+
+            _confirmed.TryAdd(block.Hash, 0);
+
+            HashDigest<SHA256>? prev = block.PreviousHash;
+            Block<T>? newTip = null;
+            do
+            {
+                if (!(prev is HashDigest<SHA256> prevHash &&
+                      Store.GetBlock<T>(prevHash) is Block<T> prevBlock))
+                {
+                    break;
+                }
+
+                uint c = _confirmed.AddOrUpdate(prevHash, k => 1U, (k, v) => v + 1U);
+                Logger.Verbose(
+                    "The block #{BlockIndex} {BlockHash} has {Confirmations} confirmations.",
+                    prevBlock.Index,
+                    prevBlock.Hash,
+                    c
+                );
+
+                if (c >= Confirmations)
+                {
+                    if (newTip is null)
+                    {
+                        newTip = prevBlock;
+                    }
+                    else
+                    {
+#pragma warning disable S1116, SA1106, SA1503
+                        // The below while statement is not "empty"; its condition has an effect.
+                        while (!_confirmed.TryRemove(prevHash, out _));
+#pragma warning restore S1116, SA1106, SA1503
+                    }
+                }
+
+                prev = prevBlock.PreviousHash;
+            }
+            while (true);
+
+            if (newTip is Block<T>)
+            {
+                Tip = newTip;
+            }
+        }
+
+        private Block<T> FindBranchpoint(Block<T> a, Block<T> b)
+        {
+            while (a is Block<T> && a.Index > b.Index && a.PreviousHash is HashDigest<SHA256> aPrev)
+            {
+                a = Store.GetBlock<T>(aPrev);
+            }
+
+            while (b is Block<T> && b.Index > a.Index && b.PreviousHash is HashDigest<SHA256> bPrev)
+            {
+                b = Store.GetBlock<T>(bPrev);
+            }
+
+            if (a is null || b is null || a.Index != b.Index)
+            {
+                throw new ArgumentException(
+                    "Some previous blocks of two blocks are orphan.",
+                    nameof(a)
```

Comment:
[`ArgumentException`](https://docs.microsoft.com/en-us/dotnet/api/system.argumentexception?view=netcore-3.1) seems designed for only one argument. 🤔

Reply:
Indeed I had thought the same question, and anyway I have no idea…